### PR TITLE
Kitsune gossip 2

### PR DESCRIPTION
**Pull request by freesig**

### CODE REVIEW

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs [hunk 1]
```diff
@@ -786,10 +793,10 @@ impl Space {
...
                         crate::gossip::sharded_gossip::recent_factory(),
                     ),
-                    (
-                        GossipModuleType::ShardedHistorical,
-                        crate::gossip::sharded_gossip::historical_factory(),
-                    ),
+                    // (
+                    //     GossipModuleType::ShardedHistorical,
+                    //     crate::gossip::sharded_gossip::historical_factory(),
+                    // ),
```
Comment (neonphog) [inline crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs:799]: Is this supposed to be commented out?

**Review by neonphog — APPROVED** [review]
Double check that commented out historical loop.

Other than that
- I like the `new_integrated_data` api addition to kitsune.
- I'm not 100% sure the complexity add of the `trigger_*` / `forced_initiates` is worth it, but as always, we can continue to iterate.

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs [hunk 2]
```diff
@@ -47,11 +47,19 @@ mod tests;
...
 type BloomFilter = bloomfilter::Bloom<Arc<MetaOpKey>>;
 type EventSender = futures::channel::mpsc::Sender<event::KitsuneP2pEvent>;
 
 struct TimedBloomFilter {
-    bloom: BloomFilter,
+    /// The bloom filter for the time window.
+    /// If this is none then we have no hashes
+    /// for this time window.
+    bloom: Option<BloomFilter>,
+    /// The time window for this bloom filter.
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:62]: Appreciate the comments

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs [hunk 3]
```diff
@@ -277,14 +278,20 @@ pub struct ShardedGossipLocalState {
     local_agents: HashSet<Arc<KitsuneAgent>>,
     tuning_params: KitsuneP2pTuningParams,
     /// If Some, we are in the process of trying to initiate gossip with this target.
-    initiate_tgt: Option<(GossipTgt, u8)>,
+    initiate_tgt: Option<(GossipTgt, u32)>,
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:281]: Even u16 is probably sufficient, but yeah, pretty minor impact either way :)

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs (hunk 1 above)
Comment (freesig) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs:799]: Whoops no

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs [hunk 4]
```diff
@@ -697,20 +764,35 @@ impl ShardedGossipLocal {
...
+                // Otherwise it's not time to sync.
                 Ok(false)
             }
         };
-        if self.have_new_data() || self.inner.share_mut(update_last_sync)? {
+
+        if self.inner.share_mut(update_last_sync)? {
             Ok(true)
         } else {
             Ok(false)
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:798]: seems like this works, no big deal though (minus the trailing brace one line down)

```suggestion
        self.inner.share_mut(update_last_sync)
```

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs [hunk 5]
```diff
@@ -277,14 +278,20 @@ pub struct ShardedGossipLocalState {
...
+    initiate_tgt: Option<(GossipTgt, u32)>,
     round_map: RoundStateMap,
     /// Metrics for a connection.
     // FIXME: Currently the p2p metric store is setup to track
     // metrics per agent but we need to track metrics per connection.
     metrics: HashMap<Tx2Cert, MetricInfo>,
     /// Last moment we locally synced.
     last_local_sync: Option<std::time::Instant>,
+    /// Trigger local sync to run on the next iteration.
+    trigger_local_sync: bool,
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:290]: This value doesn't seem to be read anywhere but maybe that's coming in future work?

**Review by maackle — APPROVED** [review]
I didn't have quite enough time before ending my day to read this thoroughly but it looks good to merge if we need to merge soon. I'll look more tomorrow if it's still open.

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs (hunk 5 above)
Comment (freesig) [reply crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:290]: It's read in the `should_local_sync` function

**Pull request merged**